pkg/pipeline/provider.go

```diff
@@ -0,0 +1,72 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2017 Datadog, Inc.
+
+package pipeline
+
+import (
+	"sync"
+
+	"github.com/DataDog/datadog-log-agent/pkg/config"
+	"github.com/DataDog/datadog-log-agent/pkg/message"
+	"github.com/DataDog/datadog-log-agent/pkg/processor"
+	"github.com/DataDog/datadog-log-agent/pkg/sender"
+)
+
+type PipelineProvider struct {
+	numberOfPipelines int
+	chanSizes         int
+	pipelinesChans    [](chan message.Message)
+
+	currentChanIdx int
+	idxMux         sync.Mutex
+}
+
+// NewPipelineProvider returns a new PipelineProvider
+func NewPipelineProvider() *PipelineProvider {
+	return &PipelineProvider{
+		numberOfPipelines: config.NumberOfPipelines,
+		chanSizes:         config.ChanSizes,
+		pipelinesChans:    [](chan message.Message){},
+		currentChanIdx:    0,
+	}
+}
+
+// Start initializes the pipelines
+func (pp *PipelineProvider) Start(cm *sender.ConnectionManager, auditorChan chan message.Message) {
+
+	for i := 0; i < pp.numberOfPipelines; i++ {
+
+		senderChan := make(chan message.Message, pp.chanSizes)
+		f := sender.New(senderChan, auditorChan, cm)
+		f.Start()
+
+		processorChan := make(chan message.Message, pp.chanSizes)
+		p := processor.New(
+			processorChan,
+			senderChan,
+			config.LogsAgent.GetString("api_key"),
+			config.LogsAgent.GetString("logset"),
+		)
+		p.Start()
+
+		pp.pipelinesChans = append(pp.pipelinesChans, processorChan)
+	}
+}
+
+func (pp *PipelineProvider) MockPipelineChans() {
+	pp.pipelinesChans = [](chan message.Message){}
+	pp.pipelinesChans = append(pp.pipelinesChans, make(chan message.Message))
+	pp.numberOfPipelines = 1
+}
+
+// Start initializes the pipelines
+func (pp *PipelineProvider) NextPipelineChan() chan message.Message {
+	pp.idxMux.Lock()
```

Comment:
While it's nothing critical, this is typically the case where an atomic would shine.

Reply:
+1